### ISSUE TITLE
Cmake error for ur_robot_driver

### ISSUE THREAD

**Issue by freesolo92** [UniversalRobots/Universal_Robots_ROS_Driver#425]
# Summary
Hey guys,

So far I have not found any clues for a solution to my problem. Now I ask for any hints.

*Introduction to the issue*

# Versions
 - ROS Driver version: ROS1 Melodic
 - Affected Robot Software Version(s): 5.10
 - Affected Robot Hardware Version(s): -
 - Robot Serial Number: -
 - UR+ product(s) installed: no
 - URCaps Software version(s): 1.0.5

# Impact
I got errors after building the sources with catkin_make_isolated. I followed the instruction in the readme file. all steps except the last one worked well.
ROS1 Melodic is successfully installed and works.

# Issue details
Output:

CMakeFiles/ur_robot_driver_node.dir/src/hardware_interface.cpp.o: In function `ur_driver::HardwareInterface::init(ros::NodeHandle&, ros::NodeHandle&)':
/home/jetson/catkin_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/src/hardware_interface.cpp:265: undefined reference to `urcl::UrDriver::UrDriver(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::function<void (bool)>, bool, std::unique_ptr<urcl::ToolCommSetup, std::default_delete<urcl::ToolCommSetup> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned int, int, double, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
collect2: error: ld returned 1 exit status
CMakeFiles/ur_robot_driver_node.dir/build.make:193: recipe for target '/home/jetson/catkin_ws/devel_isolated/ur_robot_driver/lib/ur_robot_driver/ur_robot_driver_node' failed
make[2]: *** [/home/jetson/catkin_ws/devel_isolated/ur_robot_driver/lib/ur_robot_driver/ur_robot_driver_node] Error 1
CMakeFiles/Makefile2:498: recipe for target 'CMakeFiles/ur_robot_driver_node.dir/all' failed
make[1]: *** [CMakeFiles/ur_robot_driver_node.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
<== Failed to process package 'ur_robot_driver':
  Command '['/home/jetson/catkin_ws/devel_isolated/ur_kinematics/env.sh', 'make', '-j4', '-l4']' returned non-zero exit status 2

Reproduce this error by running:
==> cd /home/jetson/catkin_ws/build_isolated/ur_robot_driver && /home/jetson/catkin_ws/devel_isolated/ur_kinematics/env.sh make -j4 -l4

Command failed, exiting.

## Use Case and Setup
This was the first try to build the driver for my Jetson Nano.

## Steps to Reproduce
When I try the same steps on my virtual machine on the pc, it works and I get no errors.
I don't know if there are any problems with ARM processors or something else.

## Expected Behavior
I expected that the build should work, following the instructions. 
But at the end of the build process, two errors occure. 

## Actual Behavior
As described above.

I am not very familiar with ROS and I am a beginner in this area, maybe someone can give me a hint, how could I fix the issue and why it occured. I thought that all dependencies should be installed, but I can be wrong.

Thanks and Greets, Marcel

**Comment by fmauch** [comment]
@freesolo92 that could be related to an installed version of the library being used for the driver. As you use `catkin_make_isolated` I assume that you also build the client library from source? In that case you could try to remove the binary installed version `sudo apt remove ros-melodic-ur-client-library`.

**Comment by freesolo92** [comment]
Hello @fmauch,

Great, it finally works!!! Thank you so much!

Does that mean that the UR Client Library is already built in ROS Melodic? 
Sorry if I ask. I don't want to waste your time, answering stupid questions. 
I have looked for an hour, if it is an bug in the /ur_robot_driver_node.dir/src/hardware_interface.cpp source file, because that was the first terminal output of the error. But no one else has had problems with the construction of the driver. So I could exclude the error.

**Comment by fmauch** [comment]
There seem to be issues that the installed version of the library shadows the one from the workspace at the linking stage.

This is definitely an issue (inside the library) but I haven't found the time to deeply look into that.

**Comment by fmauch** [comment]
As I have had trouble reproducing this in the past:

Could you please post the exact commit you are on with the ros driver and the client library inside your workspace?

`git log -1` should help with that.

**Comment by fmauch** [comment]
I found the source of the problem (I think):

As the currently released melodic version (0.2.2) has a different signature for the `UrDriver` constructor, this causes problems when the binary package is installed.

The include directories used for compiling the driver are:
```
-I/root/catkin_ws/src/Universal_Robots_ROS_Driver/ur_robot_driver/include
-I/root/catkin_ws/devel_isolated/ur_dashboard_msgs/include
-I/root/catkin_ws/src/Universal_Robots_ROS_Driver/ur_controllers/include
-I/opt/ros/melodic/include
-I/opt/ros/melodic/share/xmlrpcpp/cmake/../../../include/xmlrpcpp
-I/opt/ros/melodic/share/orocos_kdl/cmake/../../../include
-I/usr/include/eigen3 -isystem /root/catkin_ws/devel_isolated/ur_client_library/include
```

So, the header from the installed version in `/opt/ros/melodic/include` seems to get used (as CMake decided that it would be happy with the installed version). The linter, on the other hand, decides to use the version from the local workspace:
```
/usr/bin/c++  -O2 -g -DNDEBUG  -rdynamic [...] CMakeFiles/ur_robot_driver_node.dir/src/hardware_interface.cpp.o [...] /root/catkin_ws/devel_isolated/ur_client_library/lib/liburcl.so [...]
```

There, the full signatures don't match. It works, when both the header file and the library object file match, as the signature was only appended by a new entry with a default value. Therefore, the signature used inside the driver works with both the original and the modified version. Only if both are mixed, problems occur.

The good news: With the upcoming version (should be synced to the main ROS repos in a couple of days hopefully) this won't be a problem anymore. First of all, the signatures then match again, and second, there have been made improvements such that the library will be found by cmake better.

Therefore, I am going to close this as "workaround suggested, fix in release process".

**Comment by freesolo92** [comment]
I removed the whole workspace and started from a new one. After the first build of the workspace I got the same error and then tried to remove the ur-client-library as you mentioned with: `sudo apt remove ros-melodic-ur-client-library` . 
After this I tried to build the workspace again and it worked without errors.
And strangely the` git log` command does not work for me, although I had the source files from this repository.

I will read your comments in detail after my control-systems exam next week. I hope it is okay.

Okay, this means that Cmake has problems with multiply versions of a library-set (in this case the ur-client-library).

I see that it will be a long road for me to come inside the ROS world.
